server/api.go

```diff
@@ -0,0 +1,1024 @@
+	toCreate.CanCreate.Everyone = getDialogSubmissionBoolField(req, DialogFieldTypeEveryoneCanCreate)
+	toCreate.CanGrant.Everyone = getDialogSubmissionBoolField(req, DialogFieldTypeEveryoneCanGrant)
+	name, errText, errors := getDialogSubmissionTextField(req, DialogFieldTypeName)
+	if errors != nil {
+		dialogError(w, errText, errors)
+		return
+	}
+	toCreate.Name = name
+
+	createAllowList, _ := req.Submission[DialogFieldTypeAllowlistCanCreate].(string)
+	grantAllowList, _ := req.Submission[DialogFieldTypeAllowlistCanGrant].(string)
+
+	if createAllowList != "" {
+		toCreate.CanCreate.AllowList = map[string]bool{}
+		usernames := strings.Split(createAllowList, ",")
+		for _, username := range usernames {
+			username = strings.TrimSpace(username)
+			if username == "" {
+				continue
+			}
+			u, err := p.mm.User.GetByUsername(username)
+			if err != nil {
+				dialogError(w, "Cannot find user", map[string]string{DialogFieldTypeAllowlistCanCreate: fmt.Sprintf("Error getting user %s. Error: %v", username, err)})
+				return
+			}
+			toCreate.CanCreate.AllowList[u.Id] = true
+		}
+	}
+
+	if grantAllowList != "" {
+		toCreate.CanCreate.AllowList = map[string]bool{}
+		usernames := strings.Split(createAllowList, ",")
+		for _, username := range usernames {
+			username = strings.TrimSpace(username)
+			if username == "" {
+				continue
+			}
+			u, err := p.mm.User.GetByUsername(username)
+			if err != nil {
+				dialogError(w, "Cannot find user", map[string]string{DialogFieldTypeAllowlistCanGrant: fmt.Sprintf("Error getting user %s. Error: %v", username, err)})
+				return
+			}
+			toCreate.CanGrant.AllowList[u.Id] = true
+		}
+	}
+
+	_, err := p.store.AddType(toCreate)
+	if err != nil {
+		dialogError(w, err.Error(), nil)
+		return
+	}
+
+	p.mm.Post.SendEphemeralPost(userID, &model.Post{
+		UserId:    p.BotUserID,
+		ChannelId: req.ChannelId,
+		Message:   fmt.Sprintf("Type `%s` created.", toCreate.Name),
+	})
+
+	dialogOK(w)
+}
+
+// This is not working on the current webapp architecture. A similar approach should be handled using Apps.
+func (p *Plugin) dialogSelectType(w http.ResponseWriter, r *http.Request, userID string) {
+	req := model.SubmitDialogRequestFromJson(r.Body)
+	if req == nil {
+		dialogError(w, "could not get the dialog request", nil)
+		return
+	}
+
+	badgeTypeStr, errText, errors := getDialogSubmissionTextField(req, DialogFieldBadgeType)
+	if errors != nil {
+		dialogError(w, errText, errors)
+		return
+	}
+
+	badgeType, err := strconv.Atoi(badgeTypeStr)
+	if err != nil {
+		dialogError(w, "Invalid field", map[string]string{DialogFieldBadgeType: "Invalid type"})
+		return
+	}
+
+	t, err := p.store.GetType(badgesmodel.BadgeType(badgeType))
+	if err != nil {
+		dialogError(w, "Cannot get type", map[string]string{DialogFieldBadgeType: "cannot get type"})
+		return
+	}
+
+	u, err := p.mm.User.Get(userID)
+	if err != nil {
+		dialogError(w, "Cannot find user", nil)
+		return
+	}
+
+	if !canEditType(*u, *t) {
```

Comment:
Why are you passing here `*u` and `*t` instead of just receiving the pointer in the other side?